go/merkletree2/encoders_test.go

```diff
@@ -38,22 +54,96 @@ func TestBlindedSHA512_256v1Encoder(t *testing.T) {
 	require.Len(t, ks3, hashLength)
 	require.NotEqual(t, ks, ks3)
 
-	h, err = encoder.HashKeyValuePairWithKeySpecificSecret(KeyValuePair{Key: Key([]byte{0x00, 0x01}), Value: "pizza"}, ks)
+	h0, err := encoder.HashKeyEncodedValuePairWithKeySpecificSecret(KeyEncodedValuePair{Key: Key([]byte{0x00, 0x01}), Value: enc0}, ks)
 	require.NoError(t, err)
 	require.Len(t, h, hashLength)
 
-	h2, err = encoder.HashKeyValuePairWithKeySpecificSecret(KeyValuePair{Key: Key([]byte{0x00, 0x01}), Value: "pizza"}, ks)
+	h, err = encoder.HashKeyValuePairWithKeySpecificSecret(KeyValuePair{Key: Key([]byte{0x00, 0x01}), Value: val1}, ks)
+	require.NoError(t, err)
+	require.Len(t, h, hashLength)
+	require.Equal(t, h0, h)
+
+	h2, err = encoder.HashKeyValuePairWithKeySpecificSecret(KeyValuePair{Key: Key([]byte{0x00, 0x01}), Value: val1}, ks)
 	require.NoError(t, err)
 	require.Len(t, h2, hashLength)
 	require.Equal(t, h, h2)
 
-	h3, err := encoder.HashKeyValuePairWithKeySpecificSecret(KeyValuePair{Key: Key([]byte{0x00, 0x01}), Value: "pizza"}, ks3)
+	h3, err := encoder.HashKeyValuePairWithKeySpecificSecret(KeyValuePair{Key: Key([]byte{0x00, 0x01}), Value: val1}, ks3)
 	require.NoError(t, err)
 	require.Len(t, h3, hashLength)
 	require.NotEqual(t, h, h3)
 
-	h4, err := encoder.HashKeyValuePairWithKeySpecificSecret(KeyValuePair{Key: Key([]byte{0x00, 0x02}), Value: "pizza"}, ks)
+	h4, err := encoder.HashKeyValuePairWithKeySpecificSecret(KeyValuePair{Key: Key([]byte{0x00, 0x02}), Value: val1}, ks)
 	require.NoError(t, err)
 	require.Len(t, h4, hashLength)
 	require.NotEqual(t, h, h3)
 }
+
+func TestBlindedSHA512_256v1EncoderHardcodedValues(t *testing.T) {
+	// To prevent changes to go-codec or similar from breaking existing merkle
+	// proofs, this test hardcodes some example values for the encoder and
+	// hasher.
+
+	encoder := NewBlindedSHA512_256v1Encoder()
+
+	hashLength := 32
+
+	hashb1 := make([]byte, hashLength)
+	hashb1[0] = 0x01
+	hash1 := Hash(hashb1)
+
+	hashb2 := make([]byte, hashLength)
+	hashb2[0] = 0x02
+	hash2 := Hash(hashb2)
+
+	enc, err := encoder.Encode(Node{INodes: []Hash{hash1}})
+	require.NoError(t, err)
+	require.Equal(t, []byte{0x1, 0x91, 0xc4, 0x20, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, enc)
```

Comment:
In general, hardcoded values aren't great since it's unclear how to update the tests if the code changes.

Reply:
So do you suggest I remove the test? I think it is useful. I made another test above which does not hardcode values at all to catch logic errors, but this seems the only way to prevent a change in the msgpack encoding format (or in the library) from breaking proofs.
If an actual change is needed, one can always run this test, see where it fails and manually update the expected values to match the ones returned by the failing test (which is how I generated it).

Reply:
I think we should get this PR in, but it's a comment for the future. I've found myself in a position before when having to test against hardcoded vectors and not understanding what to update.